File: clients/DiscoverHosts.java
//******************************************************************************
//Copyright (c) 2012 IBM Corp.
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
//******************************************************************************

import java.util.HashMap;
import api.APIServiceClient;
import api.APIException;
import api.APINoSuchDataException;

public class DiscoverHosts {
	@SuppressWarnings("unchecked")
	public static void main(String[] args) throws APINoSuchDataException {
		 try {
			 
			 String cloudname = "TESTOPENSTACK";

			 APIServiceClient api = new APIServiceClient("172.16.1.250", 7070);

			 Object[] hostlist = (Object[]) api.perform("hostlist", cloudname);

    		 System.out.println("-----------------------------------------");
			 for (int i = 0; i < hostlist.length; i++) {
				 HashMap<String, String> hostshortdict = (HashMap<String, String>) hostlist[i];
				 String hostname = String.valueOf(hostshortdict.get("name"));
				 HashMap<String, String> hostlongdict = (HashMap<String, String>) api.perform("hostshow", cloudname, hostname);
	    		 for(String key : hostlongdict.keySet()) {
	    		     System.out.println(key + " = " + String.valueOf(hostlongdict.get(key)));
	    		  }
	    		 System.out.println("-----------------------------------------");
			 }

		 } catch (APIException e) {
		     System.err.println("APIServiceClient failure: " + e);
		 }
	}

}

File: clients/discover_hosts.py
#!/usr/bin/env python
#/*******************************************************************************
# Copyright (c) 2012 IBM Corp.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#/*******************************************************************************
'''
This is a Python example of how to discover Hosts through CloudBench

This assumes you have already attached to a cloud through the GUI or CLI.
'''

#--------------------------------- START CB API --------------------------------

from sys import path, argv
from time import sleep

import fnmatch
import os
import pwd

home = os.environ["HOME"]
username = pwd.getpwuid(os.getuid())[0]

api_file_name = "/tmp/cb_api_" + username
if os.access(api_file_name, os.F_OK) :    
    try :
        _fd = open(api_file_name, 'r')
        _api_conn_info = _fd.read()
        _fd.close()
    except :
        _msg = "Unable to open file containing API connection information "
        _msg += "(" + api_file_name + ")."
        print _msg
        exit(4)
else :
    _msg = "Unable to locate file containing API connection information "
    _msg += "(" + api_file_name + ")."
    print _msg
    exit(4)

_path_set = False

for _path, _dirs, _files in os.walk(os.path.abspath(path[0] + "/../")):
    for _filename in fnmatch.filter(_files, "code_instrumentation.py") :
        if _path.count("/lib/auxiliary") :
            path.append(_path.replace("/lib/auxiliary",''))
            _path_set = True
            break
    if _path_set :
        break

from lib.api.api_service_client import *

_msg = "Connecting to API daemon (" + _api_conn_info + ")..."
print _msg
api = APIClient(_api_conn_info)

#---------------------------------- END CB API ---------------------------------

if len(argv) < 2 :
        print "./" + argv[0] + " <cloud_name>"
        exit(1)

cloud_name = argv[1]

try :
    error = False
    hosts = None

    print "Getting hostlist on cloud \"" + cloud_name + "\"....."
    _hosts = api.hostlist(cloud_name)
    print _hosts

    for _host in _hosts :
        _host_data = api.hostshow(cloud_name, _host["name"])
        print _host_data

except APIException, obj :
    error = True
    print "API Problem (" + str(obj.status) + "): " + obj.msg

except APINoSuchMetricException, obj :
    error = True
    print "API Problem (" + str(obj.status) + "): " + obj.msg

except Exception, msg :
    error = True
    print "Problem during experiment: " + str(msg)

finally :
    if hosts is not None :
        try :
            if error :
                print "Unable to get host list"
        except APIException, obj :
            print "Error finishing up: (" + str(obj.status) + "): " + obj.msg
